### PR TITLE
[ci] Use ccache to decrease build time

### PR DESCRIPTION
**Pull request by AndrewShkrob**
Use ccache to store C++ caches for each build configuration

Stats:
| Build | Ccache | No Ccache | Increase |
|--------|--------|--------|--------|
| MacOs Debug | 4m44s | 18m41s | 3.9x |
| MacOs RelWithDebInfo | 1m8s | 23m50s | 21x |
| Linux no unity | 7m54s | 33m13s | 4.2x |
| Linux g++ Debug | 3m13s | 17m54s | 5.5x |
| Linux g++ RelWithDebInfo | 31m43s | 35m31s | 1.1x | 
| Linux clang Debug | 2m29s | 15m57s | 6.4x | 
| Linux clang RelWithDebInfo | 2m38s | 22m16s | 8.5x |

### CODE REVIEW

**Comment by vng** [comment]
Hm, is ccache really will be reused when updating PR? Or even between PRs? I thought that CI makes every build env from scratch.

Let's see, now we have 30-40 minutes. Remove the last fake commit and update this PR. Will check the new timings.

File: .github/workflows/linux-check.yaml [hunk 1]
```diff
@@ -116,6 +119,19 @@ jobs:
         shell: bash
         run: ./configure.sh
 
+      - name: ccache
+        if: github.ref != 'refs/heads/master'
+        uses: hendrikmuhs/ccache-action@v1.1
```
Comment (biodranik) [inline .github/workflows/linux-check.yaml:124]: There is already v1.2 of the action.

**Review by biodranik — COMMENTED** [review]
> A repository can have up to 10GB of caches. Once the 10GB limit is reached, older caches will be evicted based on when the cache was last accessed. Caches that are not accessed within the last week will also be evicted.

I doubt that we'll have enough space to cache for all platforms and build configs, including non-unity builds. But it may make sense to try and compare build times on PRs.

**Comment by biodranik** [comment]
Another important moment is to avoid using ccache when building for stores/production. I've heard about rare issues with ccache.

**Comment by AndrewShkrob** [comment]
@biodranik @vng PTAL

File: .github/workflows/linux-check.yaml [hunk 2]
```diff
@@ -85,6 +90,10 @@ jobs:
       matrix:
         compiler: [{ CXX: g++-12, CC: gcc-12 }, { CXX: clang++-14, CC: clang-14 }]
         CMAKE_BUILD_TYPE: [Debug, RelWithDebInfo]
+    # Cancels previous jobs if the same branch or PR was updated again.
+    concurrency:
+      group: ${{ github.workflow }}-${{ matrix.compiler.CC }}-${{ matrix.CMAKE_BUILD_TYPE }}-${{ github.event.pull_request.number || github.ref }}
```
Comment (vng) [inline .github/workflows/linux-check.yaml:95]: -unity- here? Or remove -unity- suffix below?

**Review by vng — APPROVED** [review]

**Comment by biodranik** [comment]
Linux relwithdebinfo looks like caching wasn't activated there. Is it possible to print in the action log the cache size and when it's full?

**Comment by AndrewShkrob** [comment]
It was activated. You can check stats in the `Post ccache` job.
There are a lot of cache misses. Looks like gcc RelWithDebInfo doesn't want to work with ccache
```
2023-07-09T17:47:04.6489949Z ##[group]ccache stats
2023-07-09T17:47:04.6505957Z [command]/usr/bin/ccache -s
2023-07-09T17:47:04.6992284Z Summary:
2023-07-09T17:47:04.6992886Z   Hits:              56 /  146 (38.36 %)
2023-07-09T17:47:04.6993609Z     Direct:          39 /  146 (26.71 %)
2023-07-09T17:47:04.6994059Z     Preprocessed:    17 /  107 (15.89 %)
2023-07-09T17:47:04.6994475Z   Misses:            90
2023-07-09T17:47:04.6994933Z     Direct:         107
2023-07-09T17:47:04.6996067Z     Preprocessed:    90
2023-07-09T17:47:04.6996491Z   Uncacheable:      244
2023-07-09T17:47:04.6996884Z Primary storage:
2023-07-09T17:47:04.6997326Z   Hits:             144 /  334 (43.11 %)
2023-07-09T17:47:04.6998817Z   Misses:           190
2023-07-09T17:47:04.6999310Z   Cache size (GB): 0.37 / 0.50 (73.10 %)
2023-07-09T17:47:04.6999735Z   Cleanups:          21
2023-07-09T17:47:04.7000031Z 
2023-07-09T17:47:04.7000634Z Use the -v/--verbose option for more details.
2023-07-09T17:47:04.7002045Z ##[endgroup]
2023-07-09T17:47:04.7112802Z Save cache using key "ccache-Linux Check-unity-gcc-12-RelWithDebInfo-2023-07-09T17:47:04.710Z".
2023-07-09T17:47:04.7292729Z [command]/usr/bin/tar --posix -cf cache.tzst --exclude cache.tzst -P -C /home/runner/work/organicmaps/organicmaps --files-from manifest.txt --use-compress-program zstdmt
2023-07-09T17:47:14.7115272Z Cache Size: ~348 MB (364860651 B)
2023-07-09T17:47:14.8004491Z Cache saved successfully
```

**Pull request merged**